<center>
    <h1 style="font-size:50px">
        Programmers Aren't <i style="color:red">Born</i> <br >They are <i style="color:red">Made</i> <br >
        <br >
        المبرمجين لا يولدون<br >
        انما يتم بنائهم
    </h1>
    <h4 style="font-size:20px"> 
        <i style="color:red;"> Python Programming Course </i> <br >
        <i style="color:blue;">Prepared By:  Ibrahim Muhammed Abubaker Muhammed</i>
    </h4>
</center>

# Table of Contets
* [Installing Python Language](Python Language Basics)
* [Python As Simple Calculator](Using Python For Simple Calculations)
* [Better Coding Skills](Using Jupyter Notebook and store your results in much neat way)
* [Operators](Understanding how they work)
* [Varibles](Using variable in working)
* [Datatypes](Understanding datatypes)

## Python As Simple Calculator

we can do simple python calculations and then we can put everything in this notebook so we can return to it later on..

In [9]:
# when i put # before the code it will become a comment
# a comment is unprinted code we use it to clarify for our selves the steps used in the programm
# writing so when we return after a period of time we will know exactly what we have done step by
# step

In [2]:
1+1  # addition 

2

In [3]:
2+3

5

In [3]:
5-3

2

In [4]:
2-2

0

In [5]:
2-3

-1

# OPERATORS

### Arithmetic

addition | subtraction |multiplication | division  |floor<br>division  |exponentiation | remainder
:-: | :-: | :-: | :-: | :-: | :-: | :-:
`+` | `-` |`*`| `/` | `//` | `**` | `%`

***

### Comparison
bigger than | smaller than |bigger than or equal | smaller than or equal  |not equal
 :-:        |    :-:       |     :-:             | :-:                    | :-:    
`>`         | `<`          |`<=`                 | `>=`                   | `!=`

***

### logical operators
AND   |  OR   | NOT 
:-:   |  :-:  | :-:
`and` | `or`  | `not`
 
 Bitwise AND   |  Bitwise OR          | Bitwise NOT 
 :-:           |  :-:                 | :-:
 `&`           | <code>&#124;</code>  | ``~`` 
 
 Bitwise are used binary computations and we can use them as we go along

In [8]:
5 + 5

10

In [9]:
5 * 5

25

this is exponenet

In [11]:
5**2

25

$ 6^2 $

In [4]:
6**2

36

when i divide it is better to insure numbers have $ \cdot $ in one of them

In [5]:
5/5

1

In [5]:
5.0/5.0

1.0

Why? Because There will be an **ERROR** if there is a fraction as a result, as down

In [22]:
5/2

2

In [7]:
5.0/2.0

2.5

In [6]:
23.56/4.23

5.569739952718675

In [7]:
2.5/2

1.25

a // b	Floor division	Quotient of a and b, removing fractional parts, what is left

In [8]:
5.0//2.0

2.0

In [16]:
5.0/25

0.2

In [17]:
5 % 5

0

In [8]:
7 % 2

1

a % b	Modulus	Integer remainder after division of a by b

In [18]:
5 % 2

1

In [19]:
5 / 2.0

2.5

In [20]:
5 > 6

False

In [10]:
1 >= 1

True

In [21]:
5 > 1

True

In [11]:
5 <= 1

False

In [22]:
5 >= 4

True

In [23]:
5 >= 9

False

In [24]:
5 != 1

True

In [25]:
5 != 5

False

In [53]:
False and True

False

In [56]:
False & True

False

In [57]:
True or False

True

In [58]:
True | False

True

In [59]:
False ^ True

True

In [60]:
False ^ False

False

In [61]:
not True

False


# More Math Operations

In [78]:
1 * 4 - 5 * 6

-26

It goes like <br>
$ 1 * 4 - 5 * 6 $ <br >
$ = 4 - 30 $ <br>
$ = -26 $

#### Rule Says: <br>
* First what is inside     ``()``
* Then we look at          ``* and /``
* Lastly we look at        ``+ and -`` 

In [79]:
1 * (4 - 5) * 6

-6

This Will be:<br >
$ 1 * (4 - 5) * 6 $ <br>
$ = 1 * (-1) * 6 $ <br>
$ = -6 $

In [80]:
(1 * 4) - (5 * 6)

-26

$ (1 * 4) - (5 * 6) $ <br>
$ = (4) - (30) $ <br>
$ = -26 $

In [84]:
50.0 / 4.0 - 6.0 / 7.0

11.642857142857142

In [12]:
50.0 / 4.0

12.5

In [13]:
6.0 / 7.0

0.8571428571428571

In [14]:
12.5 - 0.8571428571428571

11.642857142857142

$ 50.0 / 4.0 - 6.0 / 7.0 $ <br>
$  = 12.5 - 0.8571428571428571$ <br>
$  = 11.642857142857142 $ <br>

In [85]:
50.0 / (4.0 - 6.0) / 7

-3.5714285714285716

# Variables

x is the name of the variable we say ``x = 1`` which means we assigned the value of 1 to the ``x`` variable

In [15]:
x = 2
x**2+2*x+3

11

In [86]:
x = 1
x

1

the varibale holds it's value until we assign other value to in

In [87]:
x

1

In [11]:
x = 2
x

2

In [88]:
y = 2

In [89]:
y

2

we can add two variables togather to form some kind of logic or equation.

$x - y$

In [90]:
x - y

-1

In [91]:
x + y

3

## why we need variables
a variable is a containter we use it to save a value in order to be used later on in our program

$x^2 + 2 \cdot x + 1$

In [92]:
x = 6
x**2+2*x+1

49

$x^4 - 6 \cdot x + 49$

In [93]:
x**4-6*x+49

1309

a varibale name can be of anything as you like
<br >
$tmato = 9$
<br >
$limon = 10$

In [94]:
tomato = 9
limon = 10

we can add those varibales we have created
<br >
$
tomato + limon = 9 + 10 = 19
$

In [96]:
tomato + limon

19

But we have to be careful when we name varibels <br > 
** Erros Might Occure ** <br >
so be **careful**

In [97]:
to

NameError: name 'to' is not defined

##### Errors
when i use a variable that is not definde and not assigned to any value we get a NameError
```Python
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-97-2492535347e8> in <module>()
----> 1 to

NameError: name 'to' is not defined
```

In [98]:
lion

NameError: name 'lion' is not defined

In [99]:
tomatoo

NameError: name 'tomatoo' is not defined

##### Error
Python is case sensitive in variable calls, if you wrote a variable `tomato`, you need to call it as `tomato` otherwise you will get an error.

In [100]:
limonn

NameError: name 'limonn' is not defined

# Variable Naming Rules

In [101]:
var1 = "Ibrahim"
var1

'Ibrahim'

In [102]:
var2 = "Mohammed"

Dont Put Space When You Name A Variable It Will Cause An **ERROR**

In [103]:
var 3 = "Khalid"

SyntaxError: invalid syntax (<ipython-input-103-49cb98a12071>, line 1)

Don't Put Numbers At The Biginning of a varibale Name. <br >
Why? <br >
It Will Cause An **ERROR**

In [104]:
1var = "Ibrahim"

SyntaxError: invalid syntax (<ipython-input-104-e19f661636bc>, line 1)

**ERROR ** hyphens are not allowed

In [50]:
my-name

NameError: name 'my' is not defined

In [52]:
iam4me = "I Can Put Number In The Middle of The Name , It's Ok"
iam4me

"I Can Put Number In The Middle of The Name , It's Ok"

Varibale Names Can Be As Long As They Can Be, They Supposed To Be Clear To You And Others

In [105]:
thisIsAlongVariableName = "this is a long value"

In [106]:
thisIsAlongVariableName

'this is a long value'

In [107]:
this_is_a_long_vairable_name = "This is a very long naming"

In [108]:
this_is_a_long_vairable_name

'This is a very long naming'

# Variable Multiple Assignments

In [15]:
var1, var2, var3 = "name1",3,"five"

the results will split between variables after ther =  sign <br>
var1 will hold "name1" <br>
var2 will hold 3 <br>
var3 will hold "five"

In [16]:
var1

'name1'

In [17]:
var2

3

In [18]:
var3

'five'

In [5]:
var1 = var2 = var3 = "We Are The Same"

I can assign variables like when i say: <br>
> var1 equal var2 equal var3 equal "We Are The Same"

In [6]:
var1

'We Are The Same'

In [7]:
var2

'We Are The Same'

In [8]:
var3

'We Are The Same'

# Data Types

In [9]:
i = 19

In [10]:
i

19

In [11]:
type(i)

int

In [12]:
i = 1.6

In [13]:
i

1.6

In [14]:
type(i)

float

In [15]:
i = "Hema"

In [16]:
i

'Hema'

In [17]:
type(i)

str

In [18]:
i = "This is a long String"

In [19]:
type(i)

str

In [19]:
i = complex(1,1)
i

(1+1j)

In [20]:
type(i)

complex

In [21]:
type(True)

bool

* String is about text
* Integer is about numbers with no fraction part
* Float is about fraction and floating point
* Complex is about complex numbers and real, imaginary parts
* Boolean is about binary True and False , 1 and 0
<br>

|A                      |B
|:-:                    |:-:
|Integers               |-2, -1, 0, 1, 2, 3, 4, 5
|Floating-point numbers |-1.25, -1.0, --0.5, 0.0, 0.5, 1.0, 1.25
|Strings                |'a', 'aa', 'aaa', 'Hello!', '11 cats'
|Complex                | $ 1 + 1j , 2 - 2.0j , -6j $
|Boolean                | True , False


Note that even though the value 42 is an integer, the value 42.0 would be a floating-point number

### Why We Study data types:

In order to avoid such **ERRORS**

In [2]:
'3' + 3

TypeError: cannot concatenate 'str' and 'int' objects

In [3]:
3 + 3

6

In [4]:
'3'+'3'

'33'

### Concatination
is when you bring a string and add it to other string

In [21]:
sun_name = "Ibrahim"
father_name = "Algadi"

In [23]:
sun_name + father_name

'IbrahimAlgadi'

In [24]:
sun_name + " " + father_name

'Ibrahim Algadi'

In [25]:
one, two, three = "one", "two", "three"

In [26]:
one + " " + two + " " + three

'one two three'

### Good Concatination

In [28]:
name = sun_name + " " + father_name
name

'Ibrahim Algadi'

#### Replication Operator

In [47]:
'Ibrahim' * 5

'IbrahimIbrahimIbrahimIbrahimIbrahim'

In [48]:
'Ibrahim ' * 5

'Ibrahim Ibrahim Ibrahim Ibrahim Ibrahim '

**ERROR**  we can't multiply 2 strings togather

In [49]:
'Ibrahim' * 'Algadi'

TypeError: can't multiply sequence by non-int of type 'str'

# Help
### [Stack Overlow] (http://stackoverflow.com/) 

# Home Work
### download ** QPython **  and have python up and running in your smart phone

<center>
    <h1 style="font-size:50px">
        You Already Know What <i style="color:red">To Do</i>! <br >
        You Already Know How <i style="color:red">To Do It</i>! <br >
        So, What <i style="color:red">Stopping You</i>? <br >
        <br >
        !انت تعلم ماذا يجب ان تفعل
        <br>
        !انت تعلم كيف تقوم بفعله
        <br>
        اذا, ماذا يوقفك؟
    </h1>
</center>